File: aoc-2023/src/test/java/com/capital7software/aoc/aoc2023aoc/days/Day16Test.java
package com.capital7software.aoc.aoc2023aoc.days;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.capital7software.aoc.lib.AdventOfCodeTestBase;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

class Day16Test extends AdventOfCodeTestBase {
  private static final Logger LOGGER = LoggerFactory.getLogger(Day16Test.class);

  @BeforeEach
  void setUp() {
    var instance = new Day16();
    setupFromFile(instance.getDefaultInputFilename());
  }

  @Test
  void numberOfEnergizedTilesTopLeftHeadingRight() {
    var instance = new Day16();
    var expected = 46;
    var actual = instance.numberOfEnergizedTilesTopLeftHeadingRight(lines, false);
    assertEquals(expected, actual, "The number of energized tiles "
        + "is not what was expected: " + expected);
  }

  @Test
  void numberOfEnergizedTilesTopLeftHeadingRightRecursive() {
    var instance = new Day16();
    var expected = 46;
    var actual = instance.numberOfEnergizedTilesTopLeftHeadingRight(lines, true);
    assertEquals(expected, actual, "The number of energized tiles "
        + "is not what was expected: " + expected);
  }

  @Test
  void maximumNumberOfEnergizedTiles() {
    var instance = new Day16();
    var expected = 51;
    var actual = instance.maximumNumberOfEnergizedTiles(lines);
    assertEquals(expected, actual, "The maximum number of energized tiles "
        + "is not what was expected: " + expected);
  }

  @Override
  protected Logger getLogger() {
    return LOGGER;
  }
}

File: aoc-2024/src/test/kotlin/com/capital7software/aoc/aoc2024aoc/days/Day16Test.kt
package com.capital7software.aoc.aoc2024aoc.days

import com.capital7software.aoc.lib.AdventOfCodeTestBase
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class Day16Test : AdventOfCodeTestBase() {
  companion object {
    val log: Logger = LoggerFactory.getLogger(Day16Test::class.java)
  }

  override fun getLogger(): Logger {
    return log
  }

  @BeforeEach
  fun setup() {
    val instance = Day16()
    setupFromFile(instance.defaultInputFilename)
  }

  @Test
  fun testLowestScore() {
    val instance = Day16()
    val expected = 7036L

    val actual = instance.lowestScore(lines)
    Assertions.assertEquals(
        expected,
        actual,
        "The lowest score of $actual is not the expected lowest score of $expected."
    )
  }

  @Test
  fun testLowestScore2() {
    val instance = Day16()

    setupFromFile("inputs/input_day_16-02.txt")

    val expected = 11048L

    val actual = instance.lowestScore(lines)
    Assertions.assertEquals(
        expected,
        actual,
        "The lowest score of $actual is not the expected lowest score of $expected."
    )
  }

  @Test
  fun testBestPathTileCount() {
    val instance = Day16()
    val expected = 45L

    val actual = instance.bestPathTileCount(lines)
    Assertions.assertEquals(
        expected,
        actual,
        "The number of tiles of $actual is not the expected number of tiles of $expected."
    )
  }

  @Test
  fun testBestPathTileCount2() {
    val instance = Day16()

    setupFromFile("inputs/input_day_16-02.txt")

    val expected = 64L

    val actual = instance.bestPathTileCount(lines)
    Assertions.assertEquals(
        expected,
        actual,
        "The number of tiles of $actual is not the expected number of tiles of $expected."
    )
  }

}
